### PR TITLE
ENH: PeriodIndex can accept freq with mult

### LINKED ISSUE #7811: ENH: Allow period_range to accept freq with mult != 1
``date_range`` can accept freq with mult != 1 and output ``DatetimeIndex``, otherwise ``period_range`` raises ``ValueError``.
```
pd.date_range('2014-01-01 09:00', '2014-01-01 20:00', freq='4H')
# <class 'pandas.tseries.index.DatetimeIndex'>
# [2014-01-01 09:00:00, ..., 2014-01-01 21:00:00]
# Length: 4, Freq: 4H, Timezone: None

pd.date_range('2014-01-01 20:00', '2014-01-01 09:00', freq='-1H')
# <class 'pandas.tseries.index.DatetimeIndex'>
# [2014-01-01 20:00:00, ..., 2014-01-01 09:00:00]
# Length: 12, Freq: -1H, Timezone: None

pd.period_range('2014-01-01 09:00', '2014-01-01 20:00', freq='4H')
# ValueError: Only mult == 1 supported
```

Is it reasonable to accept ``period_range`` and ``PeriodIndex.__new__`` to this kind of freq? What I want to do is creating skipped or reversed ordered ``PeriodIndex`` easily. 
(Considering internal representation of ``PeriodIndex``, it looks better to result have normal freq,  not to add numberized-freq like ``4H``).

#### Expected
```
pd.period_range('2014-01-01 09:00', '2014-01-01 20:00', freq='4H')
# <class 'pandas.tseries.period.PeriodIndex'>
# [2014-01-01 09:00:00, ..., 2014-01-01 21:00:00]
# Length: 4, Freq: H

pd.period_range('2014-01-01 20:00', '2014-01-01 09:00', freq='-1H')
# <class 'pandas.tseries.period.PeriodIndex'>
# [2014-01-01 20:00:00, ..., 2014-01-01 09:00:00]
# Length: 12, Freq: H
```

### PR DESCRIPTION
**Pull request by sinhrks**
Closes #7811.
- [x] Change `Period.freq` and `PeriodIndex.freq` to store offsets.
  - [x] Add `freqstr` to `Period`, `PeriodIndex` can use `DatetimeIndexOpsMixin`'s logic 
  - [x] Logic and tests for pickles taken in prev versions
- [x] Perform shift/arithmetic considering freq's mult. 
  - [x] Test all offsets has accessible `n` proeprties  (#10350)
- [x] Explicit tests for `asfreq` and `to_timestamp` using freq with mult.
- [x] Ops with different base freq must be prohibited. Change freq comparison to base comparison.
- [x] Fix partial string slicing bug (some tests are commented out because of this)
- [x] Fix order bug (#10295, removing temp logic for `PeriodIndex`)
- ~~Decide a policy for legacy freq aliases, like "WK"~~ (handled in #10878)
- [x] Update doc.

### CODE REVIEW

**Comment by jreback** [comment]
@sinhrks can you rebase / revisit

**Comment by blbradley** [comment]
@jreback Indeed, the frequency given to `PeriodIndex` is not propagated to each `Period` within. Do you think it should be propagated? Your earlier response indicates that you do, but I'm checking in to verify.
This unwanted behavior is also shown using `period_range`:
```python
In [11]: pd.period_range('2-1-2014 00:00', '2-1-2014 3:00', freq='30Min').tolist()

Out[11]:

[Period('2014-02-01 00:00', 'T'),
 Period('2014-02-01 00:30', 'T'),
 Period('2014-02-01 01:00', 'T'),
 Period('2014-02-01 01:30', 'T'),
 Period('2014-02-01 02:00', 'T'),
 Period('2014-02-01 02:30', 'T'),
 Period('2014-02-01 03:00', 'T')]
```

**Comment by jreback** [comment]
@blbradley yes that sounds right. When the periods are created during iteration (e.g. the ``tolist()``) then the freq *should* (and currently is passed).

It should also be true when this PR is implementd.

**Comment by blbradley** [comment]
@jreback Thanks. I started trying to implement that. We'll see how it goes.

**Comment by sinhrks** [comment]
@jreback @blbradley OK. In #7811, I initially thought that created `PeriodIndex` doesn't need custom freq, but it is more natural. I'll consider the implementation.

**Comment by blbradley** [comment]
`Period` would greatly benefit from being rewritten in Cython (probably as part of `tslib`) as to take advantage of `convert_to_tsobject`.

**Comment by jreback** [comment]
@blbradley this is true. But that's a whole separate project (though not THAT difficult). feel free!

**Comment by blbradley** [comment]
@sinhrks I've been revisiting this. One problem is that `Period.asfreq` does not handle multiples. I think this could be handled by creating a `PeriodIndex` the same way as `DatetimeIndex` in `pandas.tseries.resample.asfreq`.

**Comment by blbradley** [comment]
@sinhrks I would love to collaborate on this. Please respond soon cause I'm actively working on it.

**Comment by jreback** [comment]
@sinhrks ?
@blbradley interested in working on this one?

**Comment by blbradley** [comment]
@jreback This year sometime! I did some code spikes towards it before, but they were not fruitful.

**Comment by sinhrks** [comment]
Maybe first change ``Period`` to take multiple freqs, then revisit this. I'll look into ``Period`` code.

**Comment by jreback** [comment]
@sinhrks can you rebase this. This would be a welcome enhancement.

**Comment by sinhrks** [comment]
Rebased. 

I've considered a little to make ``Period`` and ``PeriodIndex`` to have multiple freq, but not sure what the good behavior is.

```
import pandas as pd

pidx1 = pd.period_range('2011', freq='3A', periods=3)
pidx1
# PeriodIndex(['2011', '2014', '2017'], dtype='int64', freq='A-DEC')

# -> currently, above expr has the same as below
pd.PeriodIndex(['2011', '2014', '2017'], freq='A')

# internal representation is the same as standard frequency.
pidx1.asi8
# array([41, 44, 47])

# -> In case of multiple, it should also be continuous like array([k, k+1, k+2])?

# shift slides one freq. 
pidx1.shift(1)
# PeriodIndex(['2012', '2015', '2018'], dtype='int64', freq='A-DEC')

# -> It should slide "one multiple unit" (3 years in above case) like below?
# pidx1.shift(1)
# PeriodIndex(['2014', '2017', '2020'], dtype='int64', freq='A-DEC')
```

**Comment by jreback** [comment]
show the same example with timestamps/DatetimeIndex and see what it does, I *think* that the freqs are  multiples themselves, e.g. ``3A-DEC``

**Comment by sinhrks** [comment]
Correct. Let me organize ``DatetimeIndex`` (``TimedeltaIndex``) and current ``PeriodIndex`` behavior and ``freq`` definition differences, and would like to clarify some points.

#### Definitions

- ``DatetimeIndex`` and ``TimedeltaIndex``: Use ``freq`` to indicate whether all the values are exactly on the frequency as descent order.
- ``PeriodIndex``: Use ``freq`` to specify base frequency regardless of value order.

```
pd.DatetimeIndex(['2011-12-31', '2012-12-31', '2013-12-31'], freq='A')
# DatetimeIndex(['2011-12-31', '2012-12-31', '2013-12-31'], dtype='datetime64[ns]', freq='A-DEC', tz=None)

# Unable to instanciate DatetimeIndex in non-monotonic order
pd.DatetimeIndex(['2012-12-31', '2011-12-31', '2013-12-31'], freq='A')
# ValueError: Inferred frequency None from passed dates does notconform to passed frequency A-DEC

# But PeriodIndex can does
pd.PeriodIndex(['2012-12-31', '2011-12-31', '2013-12-31'], freq='A')
# PeriodIndex(['2011', '2013', '2012'], dtype='int64', freq='A-DEC')
```

#### Internal repl

- ``DatetimeIndex`` and ``TimedeltaIndex``: ``offsets`` instance. 
- ``PeriodIndex``: ``str``. 

```
didx = pd.DatetimeIndex(['2011-12-31', '2013-12-31', '2015-12-31'], freq='2A')
type(didx.freq)
# pandas.tseries.offsets.YearEnd

pidx = pd.PeriodIndex(['2012-12-31', '2011-12-31', '2013-12-31'], freq='A')
type(pidx.freq)
# str
```

#### Shift / Scalar arithmetic ops

- ``DatetimeIndex`` and ``TimedeltaIndex``: Shift by number of ``mult``
- ``PeriodIndex``: Shift by 1 (because ``mult`` is always 1).

```
# results in 2 years shift
didx + 1
DatetimeIndex(['2013-12-31', '2015-12-31', '2017-12-31'], dtype='datetime64[ns]', freq='2A-DEC', tz=None)
```

#### Other ops

- ``DatetimeIndex`` and ``TimedeltaIndex``: Ops with different ``base`` freq results in no ``freq``. Ops with different ``mult`` may preserve ``freq`` if possible.
- ``PeriodIndex``: Ops with different ``freq`` is prohibited. 

```
didx2 = pd.DatetimeIndex(['2012-12-31', '2013-12-31', '2014-12-31'], freq='A')
didx.union(didx2)
# DatetimeIndex(['2011-12-31', '2012-12-31', '2013-12-31', '2014-12-31', '2015-12-31'],
#               dtype='datetime64[ns]', freq='A-DEC', tz=None)

pidx2 = pd.PeriodIndex(['2012-12-31', '2011-12-31', '2013-12-31'], freq='D')
pidx.union(pidx2)
# ValueError: Only like-indexed PeriodIndexes compatible for join (for now)
```

### Required changes for ``PeriodIndex``

I think following should be basic rules?

- Internal repl: Change ``PeriodIndex.freq`` to store ``offsets``. Pickle needs to be cared.
- Shift / Arithmetic Ops: Perform ops considering ``mult``
- Other ops: Ops with different ``base`` is prohibited. Always preserve ``base`` freq, and preserve ``mult`` if possible.

**Comment by sinhrks** [comment]
OK, this should almost work now. I'd like to clarify the freq defition of ``PeriodIndex``.

As described above, ``DatetimeIndex.freq`` guarantees each timestamps are exactly on the freq.

```
import pandas as pd
pd.DatetimeIndex(['2011-12-31', '2013-12-31', '2015-12-31'], freq='2A')
# DatetimeIndex(['2011-12-31', '2013-12-31', '2015-12-31'], dtype='datetime64[ns]', freq='2A-DEC', tz=None)

pd.DatetimeIndex(['2011-12-31', '2012-12-31', '2013-12-31'], freq='2A')
# ValueError: Inferred frequency A-DEC from passed dates does not conform to passed frequency 2A-
```

Otherwise, ``PeriodIndex.freq`` only defines the base frequency of the ``Period``. Thus, following is valid.

```
pd.PeriodIndex(['2011', '2011', '2011'], freq='A')
# PeriodIndex(['2011', '2011', '2011'], dtype='int64', freq='A-DEC')
```

The point is whether ``PeriodIndex`` using freq with mult should guarantee the values are on the freq (like ``DatetimeIndex``) or not.

### 1. ``PeriodIndex`` construction using freq with mult 

Is this valid, or raise ``ValueError`` because values are not exactly on the freq with mult?

```
# current impl allows following construction.
pd.PeriodIndex(['2011', '2011', '2011'], freq='2A')
# PeriodIndex(['2011', '2011', '2011'], dtype='int64', freq='2A-DEC')
```

### 2. ``PeriodIndex`` slicing using freq with mult.

``DatetimeIndex`` multiplies its freq via slicing with step.

```
pd.DatetimeIndex(['2011-12-31', '2012-12-31', '2013-12-31'], freq='A')[::2]
# DatetimeIndex(['2011-12-31', '2013-12-31'], dtype='datetime64[ns]', freq='2A-DEC', tz=None)
```

The same rule should be applied to ``PeriodIndex``. The below case must be valid because index is monotonic.

```
# multiply freq if monotonic
pd.PeriodIndex(['2011', '2012', '2013'], freq='A')[::2]
# PeriodIndex(['2011', '2013'], dtype='int64', freq='2A-DEC')
```

In non-monotonic case, not sure whether the freq should be multiplied, or keep base freq ("A-DEC" in below example).

```
# current impl always multiplies freq.
pd.PeriodIndex(['2011', '2011', '2011'], freq='A')[::2]
# PeriodIndex(['2011', '2011'], dtype='int64', freq='2A-DEC')
```

**Comment by maximilianr** [comment]
Others will have more contextual view, but here's my $0.02, as a heavy user of `PeriodIndex`:

A `Period` is a span of time; its `freq` represents the length of that span. It exists independently of what else is in the Index (unlike `DatetimeIndex`, where the `freq` is a 'commentary' on how far the points are apart).

So
- This should be explicit in the `Period` (like it is now for quarters, as below `2011Q1`). How does the 2A example work? Is that 2011-2012? Should it clearly display that?
- I don't think you want to do any changing of the `freq` by just selecting things from an index (i.e. neither of the last two examples), because you're transforming the objects inside the index. Particularly the last example, or, even more extreme, if you had `pd.PeriodIndex(['2000', '2005', '2006'], freq='A')[::2]`, how would that work? 

Overall this is awesome and I'm excited for this to be in.

----
```
In [61]:
pd.Period('2011', freq='Q')
Out[61]:
Period('2011Q1', 'Q-DEC') 
```

**Comment by sinhrks** [comment]
> its freq represents the length of that span

Thanks. Based on the definition, it must behave like below. I'll update PR if there is no other suggestions within few days.

```
# it is valid, each elements represents the 2 years span starting from 2011 
pd.PeriodIndex(['2011', '2011', '2011'], freq='2A')
# PeriodIndex(['2011', '2011', '2011'], dtype='int64', freq='2A-DEC')

# slicing should not change freq
pd.PeriodIndex(['2011', '2012', '2013'], freq='A')[::2]
# PeriodIndex(['2011', '2013'], dtype='int64', freq='A-DEC')

# ops wish different freq is meaningless. must be prohibited 
pd.PeriodIndex(['2011', '2011', '2011'], freq='2A').union(pd.PeriodIndex(['2011', '2011', '2011'], freq='A'))
# ValueError
```

And current repl is like below:

```
pd.Period('2011', freq='2Q')
# Period('2011Q1', '2Q-DEC')
```

**Comment by maximilianr** [comment]
That's superb!

My point on the `2011Q1` example was that the definitions of the quarterly Periods are currently clear, and I wonder if the `mult` periods could be made clearer, either in the repr or the docs. 

For example, what does `pd.Period('2011', freq='2A')` represent? Is it the period from the start of 2011 until the end of 2012? 
And then, and I don't have a strong view here, should these be represented explicitly?
```
pd.Period('2011', freq='2Q') <- the period from Jan-Jun 2011 inclusive
a) # Period('2011Q1', '2Q-DEC')
b) # Period('2011Q1-Q2', '2Q-DEC')
c) # Period('2011Q1-2011Q2', '2Q-DEC')
```
```
pd.Period('2011', freq='2A') <- the period beginning in 2011, and ending in December two years later?
a) # Period('2011', '2A-DEC') 
b) # Period('2011-2012', '2A-DEC')
```
More generally, it seems like the general rule for these is 'the earliest possible valid period that starts within the date supplied'. So the first example is the first quarter of 2011, the second is the two years from the start of 2011. And `Period('2011', '2Q-NOV')` would be the Mar-May 2011 inclusive. 
Is that right?

**Comment by jreback** [comment]
@maximilianr comments are reasonable. Yes the freq is a fundamental part of the ``Period``.

When I get around to making a ``period`` dtype, it WILL have the ``freq`` as a parameter (like I am doing with the ``tz`` for a datetimes, e.g. ``datetime64[ns, US/Eastern]`` (the ``ns`` is a parameter as well, though not actually used ATM).

Notice that ``freq`` is NOT part of the dtype, rather its a property of the index.

So I would be onboard with having ``2M`` as the actual freq (stored internally prob as `base_freq`` and ``span`` maybe)

However a ``PeriodIndex`` is just a collection of ``Periods`` that happen to have the same frequency. Yes a ``DatetimeIndex`` with a freq has this same property.

Note that ``PeriodIndes`` is really just a time-based version of ``IntervalIndex`` (and maybe we can eventually change the impl to actually do that). cc @shoyer ....

**Comment by jreback** [comment]
So a period is a span, constructed from the beginning of the period. When converted it can actually convert to either the start (default) or end point.

```
In [22]: p = pd.Period('2011', freq='A')

In [23]: p
Out[23]: Period('2011', 'A-DEC')

In [24]: p.to_timestamp(how='end')
Out[24]: Timestamp('2011-12-31 00:00:00')

In [25]: p.to_timestamp(how='start')
Out[25]: Timestamp('2011-01-01 00:00:00')
```

**Comment by jreback** [comment]
@sinhrks hows this coming

**Comment by sinhrks** [comment]
Updated based on above definition, and now ready for review. Tests are locally passed.

It may better to be a separate PR, can legacy freq be deprecated?

**Comment by jreback** [comment]
@sinhrks what's the legacy freq issue?

**Comment by sinhrks** [comment]
I meant aliases defined as ``period_alias`` (defined [here](https://github.com/pydata/pandas/blob/master/pandas/tseries/frequencies.py#L648)) and [legacy aliases](http://pandas.pydata.org/pandas-docs/version/0.16.2/timeseries.html#legacy-aliases)

``period_alias``  includes some expressions such as ``WK`` which is not work in ``DatetimeIndex``. It is nice if we can choose a set of currently effective definitions. This PR doesn't touch this part yet.

```
import pandas as pd
pd.PeriodIndex(['2015-08-24'], freq='WK')
# PeriodIndex(['2015-08-24/2015-08-30'], dtype='int64', freq='W-SUN')

pd.DatetimeIndex(['2015-08-24'], freq='WK')
# ValueError: Could not evaluate WK
```

**Comment by jreback** [comment]
hmm, I don't think we have an issue to deprecate those. I'll make one; we should deprecate them in 0.17.0 and remove in next version (as they are clearly marked as legacy).

I think this PR otherwise lgtm.

**Comment by sinhrks** [comment]
Thanks, I'll prepare separate fix to deprecate legacy freqs (also going to move the legacy freq / current freq mapping to [Migrating from scikits.timeseries](http://pandas.pydata.org/pandas-docs/stable/faq.html#migrating-from-scikits-timeseries-to-pandas-0-8-0).

Let me merge this if there is no other suggestions within few days. @maximilianr, please take a look if you have a time.

**Comment by maximilianr** [comment]
@sinhrks This is awesome and I'll spend some time looking through.

First change is the docstring!
```python
Represents an period of time

Parameters
----------
value : Period or compat.string_types, default None
    The time period represented (e.g., '4Q2005')
freq : str, default None
    e.g., 'B' for businessday. Must be a singular rule-code (e.g. 5T is not
    allowed).
```

**Comment by maximilianr** [comment]
I don't think this should be happening:
```python
In [15]: p=pd.Period('2000-01',freq='3M')

In [16]: p.to_timestamp(how='S')
Out[16]: Timestamp('2000-01-01 00:00:00')

In [17]: p.to_timestamp(how='E')
Out[17]: Timestamp('2000-01-31 00:00:00')
```

**Comment by maximilianr** [comment]
I don't think this is a deal breaker, but the repr is a bit unintuitive. Here, the user has to think about what this means - I think it means a three month period from Jan-Mar 2000, but that's not immediately obvious 
(I touched on this in my comments above so if people think it's a non-issue, that's fine)

```python
Out[18]: Period('2000-01', '3M')
```

**Comment by jreback** [comment]
@maximilianr we had some discussions w.r.t. repr for ``Interval``, in https://github.com/pydata/pandas/pull/8707

I think something like this makes sense, pls open a new issue for discussions (which we can do in a future version).
- ``Period('(2000-01,2000-04]','3M')`` 
- ``Period('2000-01,2000-04','3M')`` (I think with a freq, then the ``open-close`` spans are always implicit)

**Comment by jreback** [comment]
@sinhrks lgtm. merge when ready (do the legacy issues prevent anything on this?)

**Comment by sinhrks** [comment]
@maximilianr  Thanks, good catch. I've fixed ``asfreq`` and ``to_timestamp`` as below. I've added some tests for this.

```
import pandas as pd

p = pd.Period('2000-01', freq='3M')

p.asfreq('D')      # default is how='E' (not changed)
# Period('2000-03-31', 'D')

p.asfreq('D', how='S')
# Period('2000-01-01', 'D')

p.to_timestamp()   # default is how='S' (not changed)
# Timestamp('2000-01-01 00:00:00')

p.to_timestamp(how='E')
# Timestamp('2000-03-31 00:00:00')
```

When ``mult`` is negative, it represents inverted freq starting from end. Below example means a 1 month span starting from '2000-01-31' to '2000-01-01'

```
p = pd.Period('2000-01', freq='-1M')

p.asfreq('D')
# Period('2000-01-01', 'D')

p.asfreq('D', how='S')
# Period('2000-01-31', 'D')

p.to_timestamp()
# Timestamp('2000-01-31 00:00:00')

p.to_timestamp(how='E')
# Timestamp('2000-01-01 00:00:00')
```

@jreback legacy freq can be handled separately. Let me merge this once above behavior is reviewed.

**Comment by shoyer** [comment]
What is the actual use case for negative frequencies? I think they will be a source of confusion, probably best not allowed even if they are well defined.

**Comment by jreback** [comment]
I agree about neg freqs 
unless we have a good use case lets disallow

**Comment by sinhrks** [comment]
It was defined because ``DatetimeIndex`` can do, and agreed it is meaningless in ``PeriodIndex``. Will change to raise ``ValueError``, and remove test cases.

**Comment by shoyer** [comment]
For DatetimeIndex, a negative frequency could simply correspond to a sorted index in descending order. It's not the same thing for periods.

**Comment by sinhrks** [comment]
OK, fixed to raise when ``freq`` is not positive, and added some tests. Ready for review once again.

**Comment by jorisvandenbossche** [comment]
Maybe the whatsnew notice can give it a bit more attention? (eg an example). 
Further, are there any API changes to note as well? (that freq now returns an Offset object instead of string?)

**Comment by maximilianr** [comment]
I can't spot anything else that would stop this going in.
Not sure if my comment on the docstring is wrong or it just got missed - it still states that you can't provide `5T` etc

```
    e.g., 'B' for businessday. Must be a singular rule-code (e.g. 5T is not
    allowed).
```

**Comment by sinhrks** [comment]
@jorisvandenbossche Yes, updated whatsnew and timeseries doc.

 @maximilianr I think this has been removed from ``period.pyx`` after your suggestion. If I miss other remaining, please point the line.

**Comment by maximilianr** [comment]
@sinhrks you're right - something funky was going on with my extensions - thanks

File: doc/source/whatsnew/v0.17.0.txt [hunk 1]
```diff
@@ -177,7 +177,28 @@ Other enhancements
 - ``pandas.tseries.offsets`` larger than the ``Day`` offset can now be used with with ``Series`` for addition/subtraction (:issue:`10699`).  See the :ref:`Documentation <timeseries.offsetseries>` for more details.
 
 - ``.as_blocks`` will now take a ``copy`` optional argument to return a copy of the data, default is to copy (no change in behavior from prior versions), (:issue:`9607`)
+- ``Period``, ``PeriodIndex`` and ``period_range`` can now accept multiplied freq (:issue:`7811`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.17.0.txt]: let's put this in a separate sub-section

File: doc/source/whatsnew/v0.17.0.txt [hunk 2]
```diff
@@ -177,7 +177,28 @@ Other enhancements
...
+
+  .. ipython:: python
+
+     idx = pd.period_range('2015-08-01', periods=4, freq='2D')
+     idx
+     idx + 1
+
+- ``Period.freq`` and ``PeriodIndex.freq`` are now stored as ``DateOffset`` instance, not ``str`` (:issue:`7832`)
 - ``regex`` argument to ``DataFrame.filter`` now handles numeric column names instead of raising ``ValueError`` (:issue:`10384`).
 - ``pd.read_stata`` will now read Stata 118 type files. (:issue:`9882`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.17.0.txt]: did you regenerate the legacy pickle tests pretty recently? (as I had updated maybe 2-3 weeks ago). also pls add in msgpack ones (again just regenerate, they already exist)

**Comment by jreback** [comment]
@sinhrks lgtm. couple of comments. merge after fix and green.

**Comment by sinhrks** [comment]
@jreback OK, fixed the release note and updated pickle/msgpack.

**Comment by jreback** [comment]
ok, pls rebase then merge

**Comment by sinhrks** [comment]
Rebased and now green. Merging today (Sep 3) if no further comments.

**Comment by jreback** [comment]
lgtm. merge away!

**Comment by jreback** [comment]
thanks @sinhrks awesome effort!

merging as I need to rebase tz on top of this

**Comment by sinhrks** [comment]
Thanks for review and merge:)

**Pull request merged**